pkg/monitor/ovsdb.go

```diff
@@ -160,411 +162,431 @@ func (monitor *OVSDBMonitor) startOvsdbMonitor() error {
-	monitor.cacheLock.Lock()
-	ok, localEndpointIface := monitor.isLocalEndpointPort(rowupdate.New.Fields["name"].(string))
-	if !ok {
-		monitor.cacheLock.Unlock()
-		return nil, nil
+func (monitor *OVSDBMonitor) processOvsPortDelete(uuid string, rowupdate ovsdb.RowUpdate) {
+	oldIfaces := listUUID(rowupdate.Old.Fields["interfaces"])
+	if len(oldIfaces) != 1 {
+		// bond port
+		return
 	}
-	monitor.cacheLock.Unlock()
 
-	var oldTag, newTag float64
-	var ofport uint32
-	if rowupdate.New.Fields["tag"] == nil {
-		return nil, nil
-	}
-	if rowupdate.Old.Fields["tag"] == nil {
-		return nil, nil
+	oldIfaceUUID := oldIfaces[0].GoUuid
+	oldEndpoint, ok := monitor.endpointMap[oldIfaceUUID]
+	if !ok {
+		return
 	}
-	newTag, _ = rowupdate.New.Fields["tag"].(float64)
-	oldTag, _ = rowupdate.Old.Fields["tag"].(float64)
-	if newTag == oldTag {
-		return nil, nil
+
+	if monitor.isEndpointReady(oldEndpoint) {
+		monitor.ovsdbEventHandler.DeleteLocalEndpoint(monitor.endpointMap[oldIfaceUUID])
 	}
+	delete(monitor.endpointMap, uuid)
+}
+
+func (monitor *OVSDBMonitor) processOvsInterfaceDelete(uuid string, rowupdate ovsdb.RowUpdate) {
+	// var macStr string
 
-	portName := localEndpointIface.Fields["name"].(string)
-	externalIDs := localEndpointIface.Fields["external_ids"].(ovsdb.OvsMap).GoMap
-	ofPort, ok := localEndpointIface.Fields["ofport"].(float64)
+	oldEndpoint, ok := monitor.endpointMap[uuid]
 	if !ok {
-		return nil, nil
-	}
-	if ofPort <= 0 {
-		monitor.localEndpointHardwareAddrCacheLock.Lock()
-		ofport = monitor.localEndpointHardwareAddrCache[externalIDs[LocalEndpointIdentity].(string)]
-		monitor.localEndpointHardwareAddrCacheLock.Unlock()
-	} else {
-		ofport = uint32(ofPort)
+		return
 	}
 
-	newEndpoint := monitor.interfaceToEndpoint(ofport, portName, externalIDs[LocalEndpointIdentity].(string))
-	oldEndpoint := &datapath.Endpoint{
-		InterfaceName: newEndpoint.InterfaceName,
-		MacAddrStr:    newEndpoint.MacAddrStr,
-		PortNo:        newEndpoint.PortNo,
-		BridgeName:    newEndpoint.BridgeName,
-		VlanID:        uint16(oldTag),
+	if monitor.isEndpointReady(oldEndpoint) {
+		monitor.ovsdbEventHandler.DeleteLocalEndpoint(monitor.endpointMap[uuid])
 	}
-
-	return newEndpoint, oldEndpoint
+	delete(monitor.endpointMap, uuid)
 }
 
-func (monitor *OVSDBMonitor) isLocalEndpointPort(portName string) (bool, ovsdb.Row) {
-	for i, iface := range monitor.ovsdbCache["Interface"] {
-		if iface.Fields["name"].(string) == portName {
-			if _, ok := iface.Fields["external_ids"].(ovsdb.OvsMap).GoMap[LocalEndpointIdentity]; ok {
-				return true, monitor.ovsdbCache["Interface"][i]
+func (monitor *OVSDBMonitor) getPortBridgeName(portUUID string) string {
+	var bridgeName string
+	for _, bridge := range monitor.ovsdbCache["Bridge"] {
+		portUUIDs := listUUID(bridge.Fields["ports"])
+		for _, uuid := range portUUIDs {
+			if uuid.GoUuid == portUUID {
+				bridgeName = bridge.Fields["name"].(string)
+				break
```

Comment:
break can't jump double loop, maybe you can return directly

return bridgeName

Reply:
ok